File: tests/katalon/core/KatalonDriver.js
const EventEmitter = require('events');
const moment = require('moment');
const { EventType } = require('./Constant');
const ResultResolver = require('./ResultResolver');
const TypeUtils = require('./TypeUtils');
const WebSocketClient = require('./WebsocketClient');


const colors = ['red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'white'];

class KatalonDriver {
  static index = 0;

  stepCount = 0;

  lastStepTimestamp = 0;

  // eslint-disable-next-line no-plusplus
  id = KatalonDriver.index++;

  session = new WebSocketClient();

  eventEmitter = new EventEmitter();

  resolver = ResultResolver();

  static get CURRENT_TIME() {
    return moment(new Date()).format('HH:MM:SS DD/MM/YYYY');
  }

  static get LOG_HEADER() {
    return `--- ${'[KatalonDriver]'.green}[${this.CURRENT_TIME}]`;
  }

  connect(port) {
    this.id = this.id.toString()[colors[this.id % colors.length]];
    this.session.connect(port);
    this.session.onOpen(() => {
      // console.log(`${KatalonDriver.LOG_HEADER} Connected to the Katalon RPC Server`);
    });
    this.session.onClose(() => {
      console.log(`[${this.id}] ${KatalonDriver.LOG_HEADER} Disconnected from the Katalon RPC Server`);
    });
    this.session.onError((error) => {
      console.log(`[${this.id}] ${KatalonDriver.LOG_HEADER} An error occurred in the connection with Katalon RPC Server`);
    });
    this.session.onReady(() => {
      console.log(`[${this.id}] ${KatalonDriver.LOG_HEADER} Connected to the Katalon RPC Server`);
      this.eventEmitter.emit(EventType.ready, this);
    });
  }

  log(log) {
    this.loggingSocket.emit('log', log);
  }

  close() {
    if (this.session.callbackQueue.length <= 0) {
      return this.session.close();
    }
    return setTimeout(() => {
      this.close();
    }, 1000);
  }

  onReady(listener) {
    this.eventEmitter.on(EventType.ready, listener);
  }

  call(target, method, ...args) {
    const payload = {
      object: {
        type: TypeUtils.getTypeName(target),
        id: target && target.id,
        isRemote: target && target.isRemote
      },
      method,
      args: args.map((arg) => ({
        type: TypeUtils.getTypeName(arg),
        value: TypeUtils.getRawValue(arg)
      })),
    };
    return this.session.send(JSON.stringify(payload));
  }

  set(target, prop, value) {
    const payload = {
      object: {
        type: TypeUtils.getTypeName(target),
        id: target && target.id,
        isRemote: target && target.isRemote
      },
      method: prop,
      args: [{
        type: TypeUtils.getTypeName(value),
        value: TypeUtils.getRawValue(value)
      }]
    };
    return this.session.send(JSON.stringify(payload));
  }
}


module.exports = KatalonDriver;

File: tests/Python/katalon/core/katalon_driver.py
"""
KatalonDriver
"""
import json
from eventemitter import EventEmitter
from katalon.core.websocket_client import WebSocketClient
from katalon.core.kconstants import EventType, RPCKey, RPCResponseKey
from katalon.core.type_utils import TypeUtils
from katalon.core.object_utils import Object


def map_arg(arg):
    payload_arg = {}
    payload_arg[RPCKey.type.value] = TypeUtils.get_type_name(arg)
    payload_arg[RPCKey.value.value] = TypeUtils.get_raw_value(arg)
    return payload_arg

class KatalonDriver:
    """
    KatalonDriver
    """

    def __init__(self):
        self.session = WebSocketClient()
        self.emitter = EventEmitter()
        self.stepCount = 0
        self.lastStepTimestamp = 0

    async def connect(self, port):
        self.session = WebSocketClient()
        self.session.on_ready(self.handle_ready)
        await self.session.connect(port)

    async def close(self):
        await self.session.close()

    def handle_ready(self):
        self.emitter.emit(EventType.ready, self)

    def on_ready(self, listener):
        self.emitter.on(EventType.ready, listener)

    async def call(self, target, method, *args):
        payload = {}
        payload[RPCKey.object.value] = {}
        payload[RPCKey.object.value][RPCKey.type.value] = TypeUtils.get_type_name(target)
        payload[RPCKey.method.value] = method
        payload[RPCKey.args.value] = list(map(map_arg, args))
        raw_response = await self.session.send(json.dumps(payload, ensure_ascii=False))
        data = Object.get(raw_response, RPCResponseKey.data.value)
        type_name = Object.get(raw_response, RPCResponseKey.type.value)
        return None if data is None else TypeUtils.apply_type(data, type_name)

    async def set(self, target, prop, value):
        payload = {}
        payload[RPCKey.object.value] = {}
        payload[RPCKey.object.value][RPCKey.type.value] = TypeUtils.get_type_name(target)
        payload[RPCKey.method.value] = prop
        payload[RPCKey.args.value] = list(map(map_arg, [value]))
        raw_response = await self.session.send(json.dumps(payload, ensure_ascii=False))
        return raw_response
        # return Object.get(raw_response, RPCResponseKey.data.value)

    def listen(self):
        self.session.listen()
